### PR TITLE
Move the description of the core types into their own pages

### PR DESCRIPTION
**Pull request by levitte**
This expands on some of the core type descriptions, and also makes it
easier to find the documentation for each type, at least on Unix, with
a simple call like "man OSSL_ALGORITHM".

### CODE REVIEW

File: doc/man3/OSSL_ALGORITHM.pod [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ struct ossl_algorithm_st {
+     const char *algorithm_names;     /* key */
+     const char *property_definition; /* key */
+     const OSSL_DISPATCH *implementation;
+     const char *algorithm_description;
+ };
+
+=head1 DESCRIPTION
+
+The B<OSSL_ALGORITHM> type is a I<public structure> that describe an
```
Comment (baentsch) [inline doc/man3/OSSL_ALGORITHM.pod:21]: nit: "describes"

File: doc/man3/OSSL_ALGORITHM.pod [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+=item I<algorithm_names>
+
+This string is a colon separated set of names / identities, and is used by
+the appropriate fetching functionality (such as L<EVP_CIPHER_fetch(3)>,
+L<EVP_MD_fetch(3)>, etc) to find the desired algorithm.
+
+Multiple names / identities allow a specific algorithm implementation to be
+fetched multiple ways.  For example, the RSA algorithm has the following
+know identities:
```
Comment (baentsch) [inline doc/man3/OSSL_ALGORITHM.pod:45]: nit: "known"

File: doc/man3/OSSL_ALGORITHM.pod [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+=item *
+
+C<1.2.840.113549.1.1.1>
+
+This is the OID itself for C<rsaEncryption>, in canonical decimal text form.
+
+Including this identity supports fetching OIDs that are returned by
+functions like L<OBJ_obj2txt(3)>, which is a fallback in case the OID hasn't
+been registered in F<libcrypto>'s internal object database.
```
Comment (baentsch) [inline doc/man3/OSSL_ALGORITHM.pod:71]: what about referencing the relevant registration functions here?

File: doc/man3/EVP_KDF.pod [hunk 4]
```diff
@@ -131,26 +131,26 @@ simply ignored.
...
 
 EVP_KDF_gettable_ctx_params() and EVP_KDF_CTX_gettable_params()
-return constant B<OSSL_PARAM> arrays that describe the retrievable
+return constant L<OSSL_PARAM(3)> arrays that describe the retrievable
 parameters that can be used with EVP_KDF_CTX_get_params().
 EVP_KDF_gettable_ctx_params() returns the parameters that can be retrieved
 from the algorithm, whereas EVP_KDF_CTX_gettable_params() returns
 the parameters that can be retrieved in the context's current state.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_KDF.pod:145]: Same here..

File: doc/man3/EVP_KDF.pod [hunk 5]
```diff
@@ -131,26 +131,26 @@ simply ignored.
...
 
 EVP_KDF_settable_ctx_params() and EVP_KDF_CTX_settable_params() return
-constant B<OSSL_PARAM> arrays that describe the settable parameters that
+constant L<OSSL_PARAM(3)> arrays that describe the settable parameters that
 can be used with EVP_KDF_CTX_set_params().  EVP_KDF_settable_ctx_params()
 returns the parameters that can be retrieved from the algorithm,
 whereas EVP_KDF_CTX_settable_params() returns the parameters that can
 be retrieved in the context's current state.  See L<OSSL_PARAM(3)>
-for the use of B<OSSL_PARAM> as a parameter descriptor.
+for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_KDF.pod:153]: and here..

File: doc/man3/EVP_KEYMGMT.pod [hunk 6]
```diff
@@ -88,12 +88,12 @@ of the implementations, calls I<fn> with the implementation method and
 I<data> as arguments.
 
 EVP_KEYMGMT_gettable_params() and EVP_KEYMGMT_settable_params() return a
-constant B<OSSL_PARAM> array that describes the names and types of key
+constant L<OSSL_PARAM(3)> array that describes the names and types of key
 parameters that can be retrieved or set.
 EVP_KEYMGMT_gettable_params() is used by L<EVP_PKEY_gettable_params(3)>.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_KEYMGMT.pod:94]: and here..

File: doc/man3/EVP_MAC.pod [hunk 7]
```diff
@@ -187,26 +187,26 @@ simply ignored.
 Also, what happens when a needed parameter isn't passed down is
 defined by the implementation.
 
-EVP_MAC_gettable_params() returns an B<OSSL_PARAM> array that describes
+EVP_MAC_gettable_params() returns an L<OSSL_PARAM(3)> array that describes
 the retrievable and settable parameters.  EVP_MAC_gettable_params()
 returns parameters that can be used with EVP_MAC_get_params().
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:193]: ...

File: doc/man3/EVP_MAC.pod [hunk 8]
```diff
@@ -187,26 +187,26 @@ simply ignored.
...
 
 EVP_MAC_gettable_ctx_params() and EVP_MAC_CTX_gettable_params()
-return constant B<OSSL_PARAM> arrays that describe the retrievable
+return constant L<OSSL_PARAM(3)> arrays that describe the retrievable
 parameters that can be used with EVP_MAC_CTX_get_params().
 EVP_MAC_gettable_ctx_params() returns the parameters that can be retrieved
 from the algorithm, whereas EVP_MAC_CTX_gettable_params() returns
 the parameters that can be retrieved in the context's current state.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:201]: ...

File: doc/man3/EVP_MAC.pod [hunk 9]
```diff
@@ -187,26 +187,26 @@ simply ignored.
...
 
 EVP_MAC_settable_ctx_params() and EVP_MAC_CTX_settable_params() return
-constant B<OSSL_PARAM> arrays that describe the settable parameters that
+constant L<OSSL_PARAM(3)> arrays that describe the settable parameters that
 can be used with EVP_MAC_CTX_set_params().  EVP_MAC_settable_ctx_params()
 returns the parameters that can be retrieved from the algorithm,
 whereas EVP_MAC_CTX_settable_params() returns the parameters that can
 be retrieved in the context's current state.  See L<OSSL_PARAM(3)>
-for the use of B<OSSL_PARAM> as a parameter descriptor.
+for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_MAC.pod:209]: ....

File: doc/man3/EVP_PKEY_CTX_set_params.pod [hunk 10]
```diff
@@ -30,11 +30,11 @@ These methods replace the EVP_PKEY_CTX_ctrl() mechanism. (EVP_PKEY_CTX_ctrl now
 calls these methods internally to interact with providers).
 
 EVP_PKEY_CTX_gettable_params() and EVP_PKEY_CTX_settable_params() get a
-constant B<OSSL_PARAM> array that describes the gettable and
+constant L<OSSL_PARAM(3)> array that describes the gettable and
 settable parameters for the current algorithm implementation, i.e. parameters
 that can be used with EVP_PKEY_CTX_get_params() and EVP_PKEY_CTX_set_params()
 respectively.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_PKEY_CTX_set_params.pod:37]: ...

File: doc/man3/EVP_PKEY_fromdata.pod [hunk 11]
```diff
@@ -48,10 +48,10 @@ and L<B<EVP_PKEY-ED25519>(7)|EVP_PKEY-ED25519(7)/Common X25519, X448, ED25519 an
 =for comment the awful list of links above is made this way so we get nice
 rendering as a man-page while still getting proper links in HTML
 
-EVP_PKEY_fromdata_settable() gets a constant B<OSSL_PARAM> array that describes
+EVP_PKEY_fromdata_settable() gets a constant L<OSSL_PARAM(3)> array that describes
 the settable parameters that can be used with EVP_PKEY_fromdata().
 I<selection> is described in L</Selections>.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as parameter descriptor.
```
Comment (slontis) [inline doc/man3/EVP_PKEY_fromdata.pod:54]: ...

File: doc/man3/OSSL_ALGORITHM.pod [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+Providers may find the need to register ASN.1 OIDs for algorithms using
+L<OBJ_create(3)> (via the B<core_obj_create> upcall described in
+L<provider-base(7)), because some application or library -- possibly still
+the OpenSSL libraries, even -- use NIDs to look up algorithms.
+
+In that scenario, you must make sure that the corresponding B<OSSL_ALGORITHM>'s
+I<algorithm_names> includes both the short and the long name.
+
+Most of the times, registering ASN.1 OIDs like this shouldn't be necessary,
```
Comment (slontis) [inline doc/man3/OSSL_ALGORITHM.pod:113]: time

File: doc/man7/provider-kem.pod [hunk 13]
```diff
@@ -199,10 +199,10 @@ Passing NULL for I<params> should return true.
 No parameters are currently recognised by built-in asymmetric kem algorithms.
 
 OSSL_FUNC_kem_gettable_ctx_params() and OSSL_FUNC_kem_settable_ctx_params()
-get a constant B<OSSL_PARAM> array that describes the gettable and settable
+get a constant L<OSSL_PARAM(3)> array that describes the gettable and settable
 parameters, i.e. parameters that can be used with OSSL_FUNC_kem_get_ctx_params()
 and OSSL_FUNC_kem_set_ctx_params() respectively.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as parameter descriptor.
```
Comment (slontis) [inline doc/man7/provider-kem.pod:205]: ...

File: doc/man7/provider-keyexch.pod [hunk 14]
```diff
@@ -146,17 +146,17 @@ given provider side key exchange context I<ctx> into I<params>,
...
 If OSSL_FUNC_keyexch_settable_ctx_params() is present, OSSL_FUNC_keyexch_set_ctx_params() must
 also be present, and vice versa.
-Similarly, OSSL_FUNC_keyexch_gettable_ctx_params() yields a constant B<OSSL_PARAM>
+Similarly, OSSL_FUNC_keyexch_gettable_ctx_params() yields a constant L<OSSL_PARAM(3)>
 array that describes the gettable parameters, i.e. parameters that can be
 handled by OP_signature_get_ctx_params().
 If OSSL_FUNC_keyexch_gettable_ctx_params() is present, OSSL_FUNC_keyexch_get_ctx_params() must
 also be present, and vice versa.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as parameter descriptor.
```
Comment (slontis) [inline doc/man7/provider-keyexch.pod:159]: ....

File: doc/man7/provider-signature.pod [hunk 15]
```diff
@@ -388,10 +388,10 @@ supply known values that either pass or fail.
 =back
 
 OSSL_FUNC_signature_gettable_ctx_params() and OSSL_FUNC_signature_settable_ctx_params() get a
-constant B<OSSL_PARAM> array that describes the gettable and settable parameters,
+constant L<OSSL_PARAM(3)> array that describes the gettable and settable parameters,
 i.e. parameters that can be used with OSSL_FUNC_signature_get_ctx_params() and
 OSSL_FUNC_signature_set_ctx_params() respectively.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as parameter descriptor.
```
Comment (slontis) [inline doc/man7/provider-signature.pod:394]: ....

File: doc/man3/EVP_EncryptInit.pod [hunk 16]
```diff
@@ -330,27 +330,27 @@ See L</PARAMETERS> below for more information.
...
 =item EVP_CIPHER_gettable_ctx_params() and EVP_CIPHER_CTX_gettable_params()
 
-Get a constant B<OSSL_PARAM> array that describes the retrievable parameters
+Get a constant L<OSSL_PARAM(3)> array that describes the retrievable parameters
 that can be used with EVP_CIPHER_CTX_get_params().
 EVP_CIPHER_gettable_ctx_params() returns the parameters that can be retrieved
 from the algorithm, whereas EVP_CIPHER_CTX_gettable_params() returns the
 parameters that can be retrieved in the context's current state.
-See L<OSSL_PARAM(3)> for the use of B<OSSL_PARAM> as a parameter descriptor.
+See L<OSSL_PARAM(3)> for the use of L<OSSL_PARAM(3)> as a parameter descriptor.
```
[the remark this answers is not in the document]
Comment (levitte) [reply doc/man3/EVP_EncryptInit.pod]: "Get a constant \<LINK\> ... See \<LINK\> for the use of \<LINK\> ..."

Yeah, the sanest is probably to remove that last sentence

**Pull request closed without merging**